### ISSUE TITLE
Is the nfa regex engine slow?

### ISSUE THREAD

**Issue by aktau** [neovim/neovim#650]
I'm not sure what's causing this, almost surely a plugin, but neovim freezes a lot for me. So I'm opening this issue here to get to the bottom of it when I have time. 

I tried looking at it through Xcode instruments. In these periods of heavy activity (i.e.: freezing), the nfa regex engine manages to allocate and free no less than 134GB of memory. $(deity) vim, have you never heard of a scratch buffer?!

I'm thinking the offending plugin is [vim-easytags](https://github.com/xolox/vim-easytags), which periodically runs ctags and highlights recognized symbols. This is some nice functionality that I don't want to give up, but it's triggering some pathological behaviour. Perhaps @xolox could chime in (for reference @xolox, neovim doesn't support python yet, that's going to come later). The freezing only happens whenever I move around in the buffer (which is one of the most useful things I do in vim).

Here's some screenshots of instruments measuring both CPU usage and memory, where we can see that one CPU side it's more `vim_free` having difficulty keeping up with the massive amount of free'ing going on:

![screen shot 2014-04-29 at 15 58 04](https://cloud.githubusercontent.com/assets/189413/2829704/7c0271b4-cfa6-11e3-805c-402380563d5f.png)
![screen shot 2014-04-29 at 15 51 44](https://cloud.githubusercontent.com/assets/189413/2829710/89fc0d34-cfa6-11e3-9dad-903484d52532.png)

**Comment by elmart** [comment]
Closing. May be reopened in future if profiling shows a different allocator is needed.

**Comment by justinmk** [comment]
@elmart It's ok for issues to stay open if they are still relevant. Don't be concerned with the absolute number of open issues. Also we note that we have a "reopen-later" tag.

Anyways I think this issue should stay open.

**Comment by elmart** [comment]
@justinmk Ok. I read the whole thing and my understanding was this had settled down to not doing anything unless profiling showed it was necessary. So I didn't consider it relevant anymore. Maybe we should agree on a definition for 'relevant'. Regarding the 'reopen-later' tag, I saw it, but I thought distinction between 'this-must-be-reopened' and 'this-might-be-reopened' could be useful. I think we should open a different issue to discuss this kind of things, so that I don't break your policies and we don't hijack threads with these kind of comments. Ok, I'm sorry, reopening.

**Comment by tarruda** [comment]
jemalloc is now used by default(#2415), reopening as I don't know if this issue is caused by a poor malloc implementation

**Comment by justinmk** [comment]
@aktau Are you able to reproduce this (with vim-eastytags?) after https://github.com/neovim/neovim/pull/2295 ?

**Comment by qvacua** [comment]
Just an observation I did recently (maybe already known...): Investigating qvacua/vimr#519, I observed that using `set re=1`, the old regex engine, is much faster at least for syntax highlighting. When you open the PHP file attached by @azzsh in that issue with VimR and scroll fast via trackpad, VimR stutters like crazy. Admittedly VimR's scroll performance can be improved, but it looks like that it's caused by the NFA engine being slow, since, the scrolling becomes very smooth using the old engine. `syntime report` also seems to say that the top entry is approx. 3-4 times slower with the NFA engine. Does anyone else see slow-down like this? It seem particularly bad for PHP files...

**Comment by jamessan** [comment]
Yes, the old engine is faster for certain types of regexp.  It would be worth doing some profiling with `:syntime` to find the problematic syntax definitions.

A common performance problem that was seen when the new regex engine was introduced was lookbehind assertions.  Those were enhanced to allow specifying a maximum number of bytes to look behind instead of being unbounded.

I do see some lookbehind assertions in the syntax file that may benefit from being bounded, especially if they show up as problematic in the profile.
